test/testenv/search_utils.go

```diff
@@ -0,0 +1,145 @@
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// SearchJobStatusResponse represents the search status returned by splunk for
+// endpoint: https://localhost:8089/services/search/jobs/<sid>
+type SearchJobStatusResponse struct {
+	Entries []SearchJobStatusEntry `json:"entry"`
+}
+
+// SearchJobStatusEntry represents the metadata for a given sid returned as part of the search status
+type SearchJobStatusEntry struct {
+	Name    string
+	ID      string
+	Content SearchJobStatusContent
+}
+
+// SearchJobStatusContent represents the search metadata returned as part of the search status
+type SearchJobStatusContent struct {
+	IsDone bool
+}
+
+// SearchJobResultsResponse represents the search results on non-transforming searches
+type SearchJobResultsResponse struct {
+	Fields  []SearchJobResponseFields  `json:"fields"`
+	Results []SearchJobResponseResults `json:"results"`
+}
+
+// SearchJobResponseFields represents the fields in results from non-transforming searches
+type SearchJobResponseFields struct {
+	Name string
+}
+
+// SearchJobResponseResults represents the results from non-transforming searches
+type SearchJobResponseResults struct {
+	Raw          string `json:"_raw"`
+	Source       string `json:"source"`
+	Sourcetype   string `json:"sourcetype"`
+	SplunkServer string `json:"splunk_server"`
+}
+
+// PerformSearchSync performs a syncronous search within splunk and returns the search results
+func PerformSearchSync(podName string, search string, deployment *Deployment) (string, error) {
+	// Build the search curl command and send it to an instance
+	var searchReqBuilder strings.Builder
+	curlCmd := "curl -ks -u"
+	username := "admin"
+	password := "$(cat /mnt/splunk-secrets/password) "
+	url := "https://localhost:8089/services/search/jobs/export"
+
+	fmt.Fprintf(&searchReqBuilder, "%s %s:%s %s -d output_mode=json -d search=\"search %s\"", curlCmd, username, password, url, search)
+	command := []string{"/bin/sh"}
+	searchReqResp, stderr, err := deployment.PodExecCommand(podName, command, searchReqBuilder.String(), false)
+	_ = stderr
+	if err != nil {
+		logf.Log.Error(err, "Failed to execute cmd on pod", "pod", podName, "command", command)
+		return "", err
+	}
+
+	// Since results can have multiple formats depending on the search SPL, leave this response as a string
+	return searchReqResp, err
+}
+
+// PerformSearchReq makes a search request for a search to be performed.  Returns a sid to be used to check for status and results
+func PerformSearchReq(podName string, search string, deployment *Deployment) (string, error) {
+	// Build the search curl command
+	var searchReqBuilder strings.Builder
+	curlCmd := "curl -ks -u"
+	url := "https://localhost:8089/services/search/jobs"
+	username := "admin"
+	password := "$(cat /mnt/splunk-secrets/password)"
+
+	fmt.Fprintf(&searchReqBuilder, "%s %s:%s %s -d output_mode=json -d search=\"search %s\"", curlCmd, username, password, url, search)
```

Comment:
We could but it would only be the same in the `PerformSearchReq()` and `PerformSearchSync()` functions since the others are slightly different.  Since each search/rest call will have it's own URL and POST data, I'd rather build the command here.  Though we could have a flag to switch between the async and sync versions.  But then the return values will be different.  So probably best to keep them separate, IMO.